rust/abacus-base/src/types/multisig.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use abacus_core::{MultisigSignedCheckpoint, SignedCheckpointWithSigner};
+use ethers::prelude::Address;
+use ethers::types::H256;
+
+use color_eyre::Result;
+
+use crate::{CheckpointSyncer, CheckpointSyncers};
+
+/// Fetches signed checkpoints from multiple validators to create MultisigSignedCheckpoints
+#[derive(Clone, Debug)]
+pub struct MultisigCheckpointSyncer {
+    /// The quorum threshold
+    threshold: usize,
+    /// The checkpoint syncer for each valid validator signer address
+    checkpoint_syncers: HashMap<Address, CheckpointSyncers>,
+}
+
+impl MultisigCheckpointSyncer {
+    /// Constructor
+    pub fn new(threshold: usize, checkpoint_syncers: HashMap<Address, CheckpointSyncers>) -> Self {
+        MultisigCheckpointSyncer {
+            threshold,
+            checkpoint_syncers,
+        }
+    }
+
+    /// Fetches a MultisigSignedCheckpoint if there is a quorum.
+    /// Returns Ok(None) if there is no quorum.
+    pub async fn fetch_checkpoint(&self, index: u32) -> Result<Option<MultisigSignedCheckpoint>> {
+        // Keeps track of signed validator checkpoints for a particular root.
+        // In practice, it's likely that validators will all sign the same root for a
```

Comment:
Not currently bc we don't have slashing implemented.
Atm, an improper checkpoint is only submittable in the contracts if the checkpoint has been signed by a quorum of validators. Because there's really no penalty to a single validator signing an improper checkpoint, I wanted to prevent a single validator from doing so to crash a relayer and effectively break liveness. Imo it's worth that robustness at the cost of less readability